File: 03/source.cpp

#include <fstream>
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

bool cmpr(char a, char b) {
  return a < b;
}

class Pack{
  public:
    Pack(string in) {
      int i = in.length();

      CmptA = in.substr(0,i >> 1);
      CmptB = in.substr((i >> 1),i);
      sort(CmptA.begin(), CmptA.end(), cmpr);
      sort(CmptB.begin(), CmptB.end(), cmpr);
    }

    const char Rummage() const {
      for(auto a : CmptA) {
        for(auto b: CmptB) {
          if(a == b) {
            return a;
          } else if( b > a) {
            break;
          }
        }
      }
      return 0;
    }

    const char Badges(const Pack& b, const Pack& c) const {
      string sA = CmptA + CmptB;
      string sB = b.CmptA + b.CmptB;
      string sC = c.CmptA + c.CmptB;

      sort(sA.begin(), sA.end(), cmpr);
      sort(sB.begin(), sB.end(), cmpr);
      sort(sC.begin(), sC.end(), cmpr);

      for(auto ca : sA) {
        for(auto cb : sB) {
          if(cb > ca) break;
          if( cb == ca) {
            for(auto cc : sC) {
              if(cc > cb) break;
              if(cc == cb) {
                return cc;
              }
            }
          }
        }
      }

      return 0;
    }

    string CmptA,CmptB;
};

inline static int Pri(char in) {
  return (in < 97) ? in - 38 : in - 96;
}

int main()
{
  vector<Pack> packs;

  string line;
  ifstream myfile;
  myfile.open("input.txt");

  while(getline(myfile, line)) {
    packs.push_back(Pack(line));
  }

  {// part 1
    int sum = 0;
    for(auto& p : packs) {
      const char t = p.Rummage();
      //cout << p.CmptA << " " << p.CmptB << " " << t << " " << Pri(t) << endl;
      sum += Pri(t);
    }
    cout << sum << endl;
  }

  { // part 2
    int sum = 0;
    for(int i=0; i<packs.size(); i+=3) {
      const char t = packs[i].Badges(packs[i+1], packs[i+2]);
      //cout << t << " " << Pri(t) << endl;
      sum += Pri(t);
    }
    cout << sum << endl;
  }

  return 0;
}

File: 04/source.py


def setInSet(ga,gb):
  return int(ga[0]) >= int(gb[0]) and int(ga[1]) <= int(gb[1])

def setOverlap(ga,gb):
  return (int(ga[0]) >= int(gb[0]) and int(ga[0]) <= int(gb[1]) or
          int(ga[1]) >= int(gb[0]) and int(ga[1]) <= int(gb[1]))

s1 = 0
s2 = 0
for l in open('input.txt', 'r').readlines():
  l = l.strip('\n')
  if l:
    g = l.split(',')
    ga = g[0].split('-')
    gb = g[1].split('-')

    # part 1
    if setInSet(ga,gb) or setInSet(gb,ga):
      s1 = s1 + 1
      #print(f'{ga[0]}-{ga[1]},{gb[0]}-{gb[1]} {s}')

    # part 2
    if setOverlap(ga,gb) or setInSet(gb,ga):
      s2 = s2 + 1
      #print(f'{ga[0]}-{ga[1]},{gb[0]}-{gb[1]} {s}')

print(f'Part 1: {s1}. Part 2: {s2}')
